File: leetcode-python-go/813. 最大平均值和的分组/main.go
package main

import (
	"fmt"
	"math"
)

func largestSumOfAverages(nums []int, k int) float64 {
	n := len(nums)
	prefix := make([]float64, n+1)
	for i, x := range nums {
		prefix[i+1] = prefix[i] + float64(x)
	}
	dp := make([][]float64, n+1)
	for i := range dp {
		dp[i] = make([]float64, k+1)
	}
	for i := 1; i <= n; i++ {
		dp[i][1] = prefix[i] / float64(i)
	}
	for j := 2; j <= k; j++ {
		for i := j; i <= n; i++ {
			for x := j - 1; x < i; x++ {
				dp[i][j] = math.Max(dp[i][j], dp[x][j-1]+(prefix[i]-prefix[x])/float64(i-x))
			}
		}
	}
	return dp[n][k]
}

func main() {
	nums := []int{9, 1, 2, 3, 9}
	k := 3
	results := largestSumOfAverages(nums, k)
	fmt.Println(results)
}

File: leetcode-python-go/813. 最大平均值和的分组/main.py
from typing import List
import itertools

nums = [9,1,2,3,9]
k = 3

def largestSumOfAverages(nums: List[int], k: int) -> float:
    n = len(nums)
    prefix = list(itertools.accumulate(nums, initial=0))
    dp = [[0.0] * (k + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        dp[i][1] = prefix[i] / i
    for j in range(2, k + 1):
        for i in range(j, n + 1):
            for x in range(j - 1, i):
                dp[i][j] = max(dp[i][j], dp[x][j - 1] + (prefix[i] - prefix[x]) / (i - x))
    return dp[n][k]


print(largestSumOfAverages(nums,k))